### PR TITLE
Refactorings to method/probe.rs and CrateId

### PR DESCRIPTION
**Pull request by Marwes**
A couple of refactorings done while looking into performance improvements in method resolution.

### CODE REVIEW

File: src/librustc_span/lib.rs [hunk 1]
```diff
@@ -1044,9 +1044,7 @@ impl Decodable for SourceFile {
                 name_was_remapped,
                 unmapped_path: None,
                 // `crate_of_origin` has to be set by the importer.
-                // This value matches up with `rustc_hir::def_id::INVALID_CRATE`.
```
Comment (Marwes) [inline src/librustc_span/lib.rs]: `INVALID_CRATE` appears to have been removed at some point. Figured an option would be more appropriate now.

File: src/librustc_span/lib.rs [hunk 2]
```diff
@@ -1044,9 +1044,7 @@ impl Decodable for SourceFile {
                 name_was_remapped,
                 unmapped_path: None,
                 // `crate_of_origin` has to be set by the importer.
```
Comment (petrochenkov) [inline src/librustc_span/lib.rs]: If we are using `CrateNum` we should be able to just encode and decode it here in `librustc_span/lib.rs` without help of `fn imported_source_files` in `decoder.rs`.

File: src/librustc_span/lib.rs [hunk 3]
```diff
@@ -901,7 +901,7 @@ pub struct SourceFile {
     /// Set to `None` if the `SourceFile` was imported from an external crate.
     pub unmapped_path: Option<FileName>,
     /// Indicates which crate this `SourceFile` was imported from.
-    pub crate_of_origin: u32,
+    pub crate_of_origin: Option<CrateId>,
```
[the remark this answers is not in the document]
Comment (Marwes) [reply src/librustc_span/lib.rs]: Removed that commit since it would just cause conflicts

**Pull request merged**